src/item.cpp

```diff
@@ -1516,75 +1479,226 @@ bool item::stacks_with( const item &rhs, bool check_components, bool combine_liq
+        std::set_symmetric_difference( this_mods.begin(), this_mods.end(), that_mods.begin(),
+                                       that_mods.end(), std::back_inserter( diff ) );
+        return diff.empty();
     }
-    if( max_link_length() != rhs.max_link_length() ) {
-        return false;
+    return false;
+}
+
+bool _stacks_variant( item const &lhs, item const &rhs )
+{
+    return ( !lhs.has_itype_variant() && !rhs.has_itype_variant() ) ||
+           ( lhs.has_itype_variant() && rhs.has_itype_variant() &&
+             lhs.itype_variant().id == rhs.itype_variant().id );
+}
+
+bool _stacks_components( item const &lhs, item const &rhs, bool check_components )
+{
+    return ( !check_components && !lhs.is_comestible() && !lhs.is_craft() ) ||
+           // Only check if at least one item isn't using the default recipe or is comestible
+           ( lhs.components.empty() && rhs.components.empty() ) ||
+           ( lhs.get_uncraft_components() == rhs.get_uncraft_components() );
+}
+
+} // namespace
+
+stacking_info item::stacks_with( const item &rhs, bool check_components, bool combine_liquid,
+                                 bool check_cat, int depth, int maxdepth ) const
+{
+    tname::segment_bitset bits;
+    if( type == rhs.type ) {
+        bits.set( tname::segments::TYPE );
+        bits.set( tname::segments::ENGINE_DISPLACEMENT );
+        bits.set( tname::segments::WHEEL_DIAMETER );
+        bits.set( tname::segments::WHITEBLACKLIST, _stacks_whiteblacklist( *this, rhs ) );
     }
-    const std::vector<const item *> this_mods = mods();
-    const std::vector<const item *> that_mods = rhs.mods();
-    if( this_mods.size() != that_mods.size() ) {
-        return false;
+
+    if( !check_cat && bits.none() ) {
+        return {};
     }
-    for( const item *it1 : this_mods ) {
-        bool match = false;
-        const bool i1_isnull = it1 == nullptr;
-        for( const item *it2 : that_mods ) {
-            const bool i2_isnull = it2 == nullptr;
-            if( i1_isnull != i2_isnull ) {
-                continue;
-            } else if( it1 == it2 || it1->typeId() == it2->typeId() ) {
-                match = true;
-                break;
-            }
-        }
-        if( !match ) {
-            return false;
-        }
+
+    bits.set( tname::segments::CATEGORY,
+              !check_cat ||
+              get_category_of_contents( depth, maxdepth ).get_id() ==
+              rhs.get_category_of_contents( depth, maxdepth ).get_id() );
+
+    if( check_cat && bits.none() ) {
+        return {};
     }
-    return contents.stacks_with( rhs.contents, depth, maxdepth );
+
+    bits.set( tname::segments::VARIANT, _stacks_variant( *this, rhs ) );
+    bits.set( tname::segments::RELIC, is_same_relic( rhs ) );
+
+    const std::set<ammotype> ammo = ammo_types();
+    bits.set( tname::segments::MONEY,
+              !is_money( ammo ) || ( ammo_remaining( ammo ) == rhs.ammo_remaining() ) );
```

Comment:
Probably want to check if `rhs` has the same ammo type in case it has a different item type from `this`.